### PR TITLE
PLA 2841 - API routes for generic storables

### PR DESCRIPTION
**Pull request by sfriedowitz**
# Citrine Python PR

## Description 
Through initial testing I encountered a logical need to query GEMD resources without knowing the specific type of object. It was uncovered that a handful of generic "storables" end points already exist on the product API, and PLA-2841 was open requesting functionality to hit these routes in citrine-python.

There are a handful of issues with the current draft:
- The GemResourceCollection inherits from Collection rather than DataConceptsCollection because the calls can return any type of GEMD objects/templates, rather than a single type. Further, not all methods present on DataConceptsCollection have matching end points for generic storables. It is worth thinking if this is the best inheritance structure.
- Since calls on GemResourceCollection return any type of GEMD object, we do not know the appropriate type to build the JSON response from. This is currently worked around by calling DataConcepts.get_type(data), which accesses the DataConcepts.get_type class method externally. This is not intended usage, and was under a "pragma: no cover" line. Does it make more sense for this lookup to be moved external to DataConcepts? How else to work around?

### PR Type:
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

### Adherence to team decisions
- [x] I have added tests for 100% coverage
- [x] I have written Numpy-style docstrings for every method and class.
- [x] I have communicated the downstream consequences of the PR to others.
- [x] I have bumped the version in setup.py

### CODE REVIEW

File: src/citrine/resources/gem_resource.py
Comment (bfolie) [inline src/citrine/resources/gem_resource.py:17]: I believe that this is correct

File: src/citrine/resources/gem_resource.py
Comment (bfolie) [inline src/citrine/resources/gem_resource.py:51]: Does `DataConcepts.build(data)` not work?

File: src/citrine/resources/gem_resource.py
Comment (bfolie) [inline src/citrine/resources/gem_resource.py:63]: Noooo! Don't replicate the deprecated pattern! The only argument should be `uid`, and then you call `_make_link_by_uid(uid)`.

**Comment by bfolie** [comment]
I checked the methods and they work (🎉 ).

I left a few small comments. Beyond that, the remaining work is to write tests for full coverage and to fix the flake8 style errors.

**Comment by kroenlein** [comment]
Nitpick: The resources should probably be `GemdResourceType` or `GEMDResourceType` and the file should probably be `gem_resource.py`.  GEMD refers to the data objects and GEM w/o a D (at least in this repo) is only used to refer to GEM Tables, as internal discussions showed people thought "Graphical Expression of Materials Data Tables" was lesser than "Graphical Expression of Materials Tables" (or at least as I recall).  Doesn't impact the function at all.

I also thought the storables endpoint you are hitting covered more than just data objects.

**Comment by sfriedowitz** [comment]
> Nitpick: The resources should probably be `GemdResourceType` or `GEMDResourceType` and the file should probably be `gem_resource.py`. GEMD refers to the data objects and GEM w/o a D (at least in this repo) is only used to refer to GEM Tables, as internal discussions showed people thought "Graphical Expression of Materials Data Tables" was lesser than "Graphical Expression of Materials Tables" (or at least as I recall). Doesn't impact the function at all.
> 
> I also thought the storables endpoint you are hitting covered more than just data objects.

That's a good point, I was unsure the proper naming semantics, so I followed the GEM Table example that was there already. I will go ahead and make that change.

As far as I can tell from testing, the storables end points only works for data objects (including templates). Although I haven't seen any specific documentation clarifying that.

File: tests/resources/test_gemd_resource.py
Comment (bfolie) [inline tests/resources/test_gemd_resource.py:23]: Rather than hard-coding, I suggest using random uuids (`uuid.uuid4()`) and then referring to them by variable if you need to do a comparison (though it doesn't look like any of the tests here check for the specific uuids)

**Comment by bfolie** [comment]
Good catch about GEM vs. GEMD @kroenlein.

@sfriedowitz, when Ken says "only GEMD objects" I believe he means "not templates." We all expect the `storables` endpoint to work for objects and templates, which is why we went with the name `GEMDResourceCollection`, since "resource" seemed general enough to cover both objects and templates. Ken, is there someplace in the PR where it is described as working specifically for data objects?

File: tests/resources/test_gemd_resource.py
Comment (sfriedowitz) [reply tests/resources/test_gemd_resource.py]: Changed in most recent push, along with a few doc updates.

**Review by bfolie — DISMISSED** [review]
Nice job. To close the loop, I suggest tinkering with our end-to-end tests a bit so that the storables route is exercised somewhere

**Comment by sfriedowitz** [comment]
@bfolie Slight change of plan. Just talked with Hutch, and we agree it makes most sense to change the inheritance chain for GEMDResource to derive from DataConceptsCollection. Will edit and post again for review.

File: tests/resources/test_gemd_resource.py
Comment (kroenlein) [inline tests/resources/test_gemd_resource.py:48]: Using string formatting methods to assemble URLs isn't great practice (see #650), but it's not worth sweating since this is only a test.  Any reason you didn't use an `f""` string?
```suggestion
        path=format_escaped_url('projects/{}/storables', collection.project_id),
```

File: src/citrine/resources/gemd_resource.py
Comment (kroenlein) [inline src/citrine/resources/gemd_resource.py:53]: Are we forbidding the query `dataset.storables.update(...)`?  It seems like these should check if `self.dataset_id` is set, and dispatch if possible.  Regardless, I don't think this is a `NotImplementedError` error, this would probably be a `ValueError` since they tried to invoke it with `dataset is None`.

https://docs.python.org/3/library/exceptions.html#NotImplementedError

File: src/citrine/resources/gemd_resource.py
Comment (kroenlein) [inline src/citrine/resources/gemd_resource.py:12]: The most consistent class name is probably `GEMDEntity` since that's the parent element of both object and template in gemd-python, though that includes several classes that are not directly supported with this class.

**Review by kroenlein — COMMENTED** [review]
A few more nits.  I started a bunch of comments a few days ago to support consistency w/ #650, but they got obviated with the inheritance from DataConcepts (I think).

File: tests/resources/test_gemd_resource.py
Comment (sfriedowitz) [reply tests/resources/test_gemd_resource.py:48]: easy fix, thanks for pointing out. and yes, all implementations were passed to DataConceptsCollection after switching inheritance, so should be consistent.

File: src/citrine/resources/gemd_resource.py
Comment (sfriedowitz) [reply src/citrine/resources/gemd_resource.py:53]: @bfolie Can use your input here. The logic was that since the storables route does not have functionality for register/update/delete, we would prefer the user to use the methods on specific collections or on `dataset`, rather than dispatching to that through `GEMDCollection`. 

We can add that dispatch through the parent dataset, but it seems redundant to me. Think it's worth adding in?

File: src/citrine/resources/gemd_resource.py
Comment (bfolie) [reply src/citrine/resources/gemd_resource.py:12]: I agree that "Entity" is the most appropriate term, if we're inheriting terminology from `gemd-python`. Doesn't it include everything? It looks like the things in `gemd-python` that inherit `BaseEntity` are just objects, object templates, and attribute templates.

The term "Entity" is used in some places in `citrine-python`, mostly in types as part of docstrings, so I wouldn't be opposed to calling this `GEMDEntityCollection`. On the other hand, `citrine-python` heavily uses the term "Resource" to refer to objects that are stored on-platform. The directory is `/resources`, and we also talk about "AI Resources" in places.

File: src/citrine/resources/data_concepts.py
Comment (bfolie) [inline src/citrine/resources/data_concepts.py]: Can't this be a template as well?

File: src/citrine/resources/gemd_resource.py
Comment (bfolie) [reply src/citrine/resources/gemd_resource.py:53]: If the dataset is not None, then it's a `NotImplementedError`, because we haven't implemented it :p
If the method is otherwise implemented but the dataset is `None`, then we've been using `RuntimeError`.

register/update/delete is a nice-to-have. I was more ambivalent when we last discussed it, but now I think it's worth doing, both because users will expect it and because it should be easy to implement by calling up to the already-implemented `Dataset` method. It's true that the `storables/` route doesn't explicitly implement, say, a POST method, but that's immaterial to a user. As far as they're concerned, their intent is clear, so we should be able to deal with it.

File: src/citrine/resources/gemd_resource.py
Comment (bfolie) [inline src/citrine/resources/gemd_resource.py:92]: This is an internal method, so we don't need to override it

File: src/citrine/resources/gemd_resource.py
Comment (sfriedowitz) [reply src/citrine/resources/gemd_resource.py:12]: I'm falling back to `resource` being the most fitting term within `citrine-python`. It is used significantly throughout the other collections, and seems general enough. I will opt to keep as such for this iteration.

**Comment by sfriedowitz** [comment]
New storables additions to the product API will be soon be added to handle generic register/update/delete calls. This will require an implementation change of the current iteration. Tabling as a draft until then.

**Comment by sfriedowitz** [comment]
The full interface is fleshed out on GEMDResourceCollection. This involves a refactor of dataset methods and associated tests to live on GEMDResourceCollection, as well as a lookup method `DataConcepts.get_collection_type` to determine the associated collection type for a given object.

One main question: Do we want to deprecate calls to `dataset.register`, `dataset.update`, etc.., since these generic methods now live on the generic GEMDResourceCollection, or do we want a subset of these methods to remain on Dataset?

**Comment by bfolie** [comment]
> One main question: Do we want to deprecate calls to `dataset.register`, `dataset.update`, etc.., since these generic methods now live on the generic GEMDResourceCollection, or do we want a subset of these methods to remain on Dataset?

I don't think that we need to deprecate those methods. It doesn't hurt to give users more idiomatic ways to do the same thing, as long as the intent is clear.

**Comment by bfolie** [comment]
Because this is a significant change and involves API routes, I suggest you run the end-to-end tests using this branch and we make sure that they pass before we merge. Do you know how to do that?

**Comment by sfriedowitz** [comment]
> Because this is a significant change and involves API routes, I suggest you run the end-to-end tests using this branch and we make sure that they pass before we merge. Do you know how to do that?

File: src/citrine/resources/gemd_resource.py
Comment (kroenlein) [inline src/citrine/resources/gemd_resource.py:24]: Why are these defined in this class?  The unit tests pass with these commented out.

File: src/citrine/resources/gemd_resource.py
Comment (kroenlein) [inline src/citrine/resources/gemd_resource.py:185]: This method doesn't grammatically make sense as the gemd-collection; it's supposed to delete the contents of a dataset, so the method should probably continue living there.

**Review by kroenlein — COMMENTED** [review]
A couple questions.

File: src/citrine/resources/gemd_resource.py
Comment (sfriedowitz) [reply src/citrine/resources/gemd_resource.py]: > This method doesn't grammatically make sense as the gemd-collection; it's supposed to delete the contents of a dataset, so the method should probably continue living there.

Makes sense, changed it back!

File: src/citrine/resources/gemd_resource.py
Comment (sfriedowitz) [reply src/citrine/resources/gemd_resource.py]: Yes they appear unused, so am removing in the changes.

**Comment by sfriedowitz** [comment]
> Because this is a significant change and involves API routes, I suggest you run the end-to-end tests using this branch and we make sure that they pass before we merge. Do you know how to do that?

ITs passing with current build (https://jenkins.corp.citrine.io/job/nextgen-ITs/18296/). Also planning to add some test coverage to nextgen specific to these routes.

**Comment by sfriedowitz** [comment]
> > Because this is a significant change and involves API routes, I suggest you run the end-to-end tests using this branch and we make sure that they pass before we merge. Do you know how to do that?
> 
> ITs passing with current build (https://jenkins.corp.citrine.io/job/nextgen-ITs/18296/). Also planning to add some test coverage to nextgen specific to these routes.

And a test on nextgen for when it goes live: https://github.com/CitrineInformatics/nextgen-devkit/tree/PLA-2841/generic-storables

File: src/citrine/resources/gemd_resource.py
Comment (bfolie) [reply src/citrine/resources/gemd_resource.py]: They're not used in this case, but they are part of the interface, so I would vote for retaining them.

But I won't block on it, and am approving.

**Review by bfolie — APPROVED** [review]

**Pull request merged**
